### ISSUE TITLE
Multiple Storable Objects

### ISSUE THREAD

**Issue by crisward** [kemalcr/kemal-session#37]
If I set multiple storable objects, the object type seems to get mismatched. This seems to happen when the data gets read back from a file, and it parsed with .from_json.

Let me know if you need more info to recreate.

**Comment by neovintage** [comment]
@crisward can you post an example? That would help me a ton.

**Comment by crisward** [comment]
Will do. I've worked around this at the moment by creating a session schema object and storing everything in there as per this discussion - https://github.com/kemalcr/kemal-session/issues/38. 

I'll try and add a failing test...

**Comment by neovintage** [comment]
@crisward what version of LLVM do you have installed?

**Comment by neovintage** [comment]
Alright, I think I figured it out. The specs as they exist today between `UserTestSerialization` and `UserTestDeserialization` will result in type mismatches when parsing the json from the storage instance. The reason this happens is that when a union type is deserialized from JSON, crystal will go through all the types within the union until one of them parses correctly. Unless the converter method throws an error, it's likely that you'll get a parse mismatch on deserialization. If you have different classes but they happen to have the same properties, like in cases where you subclass, you're always going to get the incorrect type. 

Our options include:
1. include the _actual_ type when serializing the Storable Object to session storage.
2. Only allow one storable object across a code base

I'm in favor of the first option. I think that's going to mean we're going to have to create a wrapper object to allow us to encode the type as part of the storable object. I don't think that will require the storage engines to have to change their implementations but I'm not sure at this point.

I think there's a way forward but I'm not sure when I can get some time to fix it.

**Comment by crisward** [comment]
I went with option no.2 in my usage and just have a big 'session' object which I've attached my other objects to, as per this discussion - https://github.com/kemalcr/kemal-session/issues/38. Sorry I never did get around to giving an example. It was a case of finding a solution and moving on. I intend to write a mysql implementation of session storage once these kind of implementation details are ironed out.
